File: lighting.py
# src/iluminacao.py
import numpy as np
from src.Ponto import Ponto
from src.Vetor import Vetor


# ============================================================
# CONSTANTES
# ============================================================
EPSILON = 1e-4


# ============================================================
# SOMBRAS
# ============================================================
def checar_sombra(
    ponto_deslocado: Ponto,
    L_normalizado: Vetor,
    distancia_luz: float,
    objetos: list,
    intersect_func
) -> bool:
    """
    Dispara um shadow ray a partir de um ponto ligeiramente deslocado.
    
    Retorna True se existir qualquer objeto entre:
        0 < t < distancia_luz
    """

    for obj_sombra in objetos:
        t_sombra, _ = intersect_func(
            obj_sombra,
            ponto_deslocado,
            L_normalizado
        )

        if EPSILON < t_sombra < distancia_luz:
            return True

    return False


# ============================================================
# PHONG
# ============================================================
def calcular_cor_phong(
    P: Ponto,
    N: Vetor,
    ray_dir: Vetor,
    hit_obj,
    scene_data,
    intersect_func
) -> tuple[float, float, float]:

    Ia = scene_data.global_light.color
    mat = hit_obj.material

    # =========================================================
    # NORMALIZAÇÃO GLOBAL
    # =========================================================

    N = N.normalize()
    #ray_dir = ray_dir.normalize()

    # Corrige normal para ficar contra o raio incidente
    '''if ray_dir.dot(N) > 0:
        N = N * (-1.0)'''

    # Vetor de visão
    V = (scene_data.camera.lookfrom - P).normalize()

    # =========================================================
    # COMPONENTE AMBIENTE
    # =========================================================

    cor_r = mat.ka.r * Ia.r
    cor_g = mat.ka.g * Ia.g
    cor_b = mat.ka.b * Ia.b

    # =========================================================
    # LUZES
    # =========================================================

    for luz in scene_data.light_list:

        vetor_luz = luz.pos - P

        distancia_luz = vetor_luz.magnitude()

        # Vetor unitário direção da luz
        L = vetor_luz.normalize()

        # =====================================================
        # SHADOW RAY
        # =====================================================

        P_sombra = P + (N * EPSILON)

        if checar_sombra(
            P_sombra,
            L,
            distancia_luz,
            scene_data.objects,
            intersect_func
        ):
            continue

        # =====================================================
        # DIFUSA
        # =====================================================

        L_dot_N = N.dot(L)

        if L_dot_N <= 0:
            continue

        cor_r += (
            mat.color.r *
            L_dot_N *
            luz.color.r
        )

        cor_g += (
            mat.color.g *
            L_dot_N *
            luz.color.g
        )

        cor_b += (
            mat.color.b *
            L_dot_N *
            luz.color.b
        )

        # =====================================================
        # ESPECULAR
        # =====================================================

        # R = 2(N·L)N - L

        R = ((N * (2.0 * L_dot_N)) - L).normalize()

        R_dot_V = R.dot(V)

        if R_dot_V > 0:

            spec = R_dot_V ** mat.ns

            cor_r += (
                mat.ks.r *
                spec *
                luz.color.r
            )

            cor_g += (
                mat.ks.g *
                spec *
                luz.color.g
            )

            cor_b += (
                mat.ks.b *
                spec *
                luz.color.b
            )

    # =========================================================
    # CLAMP
    # =========================================================

    cor_r = min(1.0, max(0.0, cor_r))
    cor_g = min(1.0, max(0.0, cor_g))
    cor_b = min(1.0, max(0.0, cor_b))

    return cor_r, cor_g, cor_b

File: core/lighting.cpp
#include "lighting.h"
#include <cmath>
#include <algorithm>

const double EPSILON = 1e-4;

namespace {
    bool checar_sombra(
        const Ponto& ponto_deslocado,
        const Vetor& L_normalizado,
        double distancia_luz,
        const std::vector<ObjectData>& objetos,
        const std::function<HitResult(const ObjectData&, const Ponto&, const Vetor&)>& intersect_func
    ) {
        for (const auto& obj_sombra : objetos) {
            HitResult hr = intersect_func(obj_sombra, ponto_deslocado, L_normalizado);
            if (hr.hit() && hr.t > EPSILON && hr.t < distancia_luz) {
                return true;
            }
        }
        return false;
    }
}

std::array<double, 3> calcular_cor_phong(
    const Ponto& P,
    const Vetor& N_in,
    const Vetor& ray_dir,
    const ObjectData& hit_obj,
    const ColorData& cor_global,
    const Ponto& LookFrom_atual,
    const std::vector<LightData>& LightList,
    const std::vector<ObjectData>& valid_objects,
    const std::function<HitResult(const ObjectData&, const Ponto&, const Vetor&)>& intersect_func
) {
    auto Ia = cor_global;
    auto mat = hit_obj.material;

    Vetor N = N_in.normalize();
    Vetor V = (-ray_dir).normalize();

    double cor_r = mat.ka.r * Ia.r;
    double cor_g = mat.ka.g * Ia.g;
    double cor_b = mat.ka.b * Ia.b;

    for (const auto& luz : LightList) {
        Vetor vetor_luz = luz.pos - P;
        double distancia_luz = vetor_luz.magnitude();
        Vetor L = vetor_luz.normalize();

        // Atenuação simples: 1.0 / (1.0 + 0.1 * distancia)
        // Isso evita o estouro de luz sem adicionar novas dependências complexas
        double atenuacao = 1.0 / (1.0 + 0.1 * distancia_luz);

        double L_dot_N = N.dot(L);
        if (L_dot_N <= 0.0) continue;

        Ponto P_sombra = P + (N * EPSILON);
        if (checar_sombra(P_sombra, L, distancia_luz, valid_objects, intersect_func)) {
            continue;
        }

        // Difuso com atenuação
        cor_r += mat.color.r * L_dot_N * luz.color.r * atenuacao;
        cor_g += mat.color.g * L_dot_N * luz.color.g * atenuacao;
        cor_b += mat.color.b * L_dot_N * luz.color.b * atenuacao;

        // Especular com atenuação
        Vetor R = ((N * (2.0 * L_dot_N)) - L).normalize();
        double R_dot_V = R.dot(V);

        if (R_dot_V > 0.0) {
            double spec = std::pow(R_dot_V, mat.ns);
            cor_r += mat.ks.r * spec * luz.color.r * atenuacao;
            cor_g += mat.ks.g * spec * luz.color.g * atenuacao;
            cor_b += mat.ks.b * spec * luz.color.b * atenuacao;
        }
    }

    cor_r = std::clamp(cor_r, 0.0, 1.0);
    cor_g = std::clamp(cor_g, 0.0, 1.0);
    cor_b = std::clamp(cor_b, 0.0, 1.0);

    return {cor_r, cor_g, cor_b};
}